deku-core/src/Deku/Internal/Region.purs

```diff
@@ -0,0 +1,536 @@
+    ownerIx <- ownerEff
+
+    extentToEff <- ST.read prevExtent.extent
+    extentToIx <- extentToEff
+    selfIx <- bumped.ix
+
+    previousRegion <- runSTFn2 index (selfIx - 1) children
+
+    -- find the smallest possible update, but never update the first element, the parent
+    if selfIx - ownerIx > extentToIx - selfIx || ownerIx == 0 then do
+      -- the following `SharedBound` will be smaller
+      newShared <- do
+        owner <- ST.new bumped.ix
+        extent <- ST.new extentToEff
+        ref <- ST.new anchor
+        let
+          pushAnchor = mkSTFn1 \a -> void $ ST.write a ref 
+          bound = ST.read ref
+        pure { owner, extent, bound, pushAnchor }
+
+      runSTFn4 fixManagedTo selfIx ( extentToIx + 1 ) (updateShared newShared)
+        children
+        
+      -- update the extent of the preceding `SharedBound` to the previous `ManagedRegion`
+      void $ ST.write previousRegion.ix prevExtent.extent
+
+    else do
+      -- preceding `SharedBound` will be smaller
+      newShared <- do
+        owner <- ST.new ownerEff
+        extent <- ST.new previousRegion.ix
+        ref <- ST.new =<< prevExtent.bound
+        let
+          pushAnchor = mkSTFn1 \a -> void $ ST.write a ref 
+          bound = ST.read ref
+        pure { owner, extent, bound, pushAnchor }
+
+      runSTFn4 fixManagedTo ownerIx selfIx (updateShared newShared)
+        children
+
+      -- hijack longer prevExtent `SharedBound` and update it with our own info, making bumped the new owner
+      void $ ST.write bumped.ix prevExtent.owner
+      runSTFn1 prevExtent.pushAnchor anchor
+
+insertManaged
+  :: STFn2 (Maybe Int) Children Global
+       ManagedRegion
+insertManaged = mkSTFn2 \givenPos children -> do
+  length <- STArray.length children
+  let
+    -- clamp and correct the givenPos, because the parent `SharedBound` already occupies the first element and shifts
+    -- everything else by one
+    pos :: Int
+    pos =
+      clamp 1 length $ maybe length (add 1) givenPos
+
+  ixRef <- ST.new pos
+  posEvent <- createPure
+
+  let
+    pushIx :: STFn1 Int Global Unit
+    pushIx = mkSTFn1 \i -> do
+      void $ ST.write i ixRef
+      posEvent.push $ i - 1 -- and restore position for the user
+
+    ix :: ST.ST Global Int
+    ix =
+      ST.read ixRef
+
+    position :: Poll Int
+    position = stRefToPoll ixRef <|> pollFromEvent posEvent.event
+
+  prevBegin <- runSTFn2 shareBound ix children
+  end <- ST.new prevBegin
+
+  let
+    managed :: ManagedRegion
+    managed =
+      { ix, pushIx, position, end }
+
+  void $ STArray.splice pos 0 [ managed ] children
+  runSTFn3 fixManaged (pos + 1) updateIx children
+  pure managed
+
+updateShared :: SharedBound' -> STFn2 Int ManagedRegion Global Unit
+updateShared shared = mkSTFn2 \_ region -> do
+  void $ ST.write shared region.end
+
+updateIx :: STFn2 Int ManagedRegion Global Unit
+updateIx =
+  mkSTFn2 \i { pushIx } -> runSTFn1 pushIx i
+
+-- TODO: expensive operation
```

Comment:
Is it possible to defer this operation until it's needed? If so, you could try to make it lazy. This is of course where Haskell shines, but there are ways to build little runtimes that'll manage this sort of thing for you - lazy lists do that, for example. Of course, laziness in JS leads to thunks, so it's only worth it if the thunk tax is lower than the cost of execution.

Reply:
A lot of stuff relies on the index information. I can probably reduce this a bit but then again the read to write ratio for that is kinda bad. I.e. basically every operation uses it(for now) and it's only expensive when using `sentTo` or inserting at the beginning.
It's a problem but I would wait for it to show up in the benchmarks.